pkg/clients/eks/nodegroup.go

```diff
@@ -0,0 +1,202 @@
+// from eks.Nodegroup.
+func GenerateNodeGroupObservation(ng *eks.Nodegroup) v1alpha1.NodeGroupObservation { // nolint:gocyclo
+	if ng == nil {
+		return v1alpha1.NodeGroupObservation{}
+	}
+	o := v1alpha1.NodeGroupObservation{
+		NodeGroupArn: awsclients.StringValue(ng.NodegroupArn),
+		Status:       v1alpha1.NodeGroupStatusType(ng.Status),
+	}
+	if ng.CreatedAt != nil {
+		o.CreatedAt = &metav1.Time{Time: *ng.CreatedAt}
+	}
+	if ng.Health != nil && len(ng.Health.Issues) > 0 {
+		o.Health = v1alpha1.NodeGroupHealth{
+			Issues: make([]v1alpha1.Issue, len(ng.Health.Issues)),
+		}
+		for c, i := range ng.Health.Issues {
+			o.Health.Issues[c] = v1alpha1.Issue{
+				Code:        v1alpha1.NodeGroupIssueCode(i.Code),
+				Message:     aws.StringValue(i.Message),
+				ResourceIDs: i.ResourceIds,
+			}
+		}
+	}
+	if ng.ModifiedAt != nil {
+		o.CreatedAt = &metav1.Time{Time: *ng.ModifiedAt}
+	}
+	if ng.Resources != nil {
+		o.Resources = v1alpha1.NodeGroupResources{
+			RemoteAccessSecurityGroup: aws.StringValue(ng.Resources.RemoteAccessSecurityGroup),
+		}
+		if len(ng.Resources.AutoScalingGroups) > 0 {
+			asg := make([]v1alpha1.AutoScalingGroup, len(ng.Resources.AutoScalingGroups))
+			for c, a := range ng.Resources.AutoScalingGroups {
+				asg[c] = v1alpha1.AutoScalingGroup{Name: aws.StringValue(a.Name)}
+			}
+			o.Resources.AutoScalingGroups = asg
+		}
+	}
+	return o
+}
+
+// LateInitializeNodeGroup fills the empty fields in *v1alpha1.NodeGroupParameters with the
+// values seen in eks.Nodegroup.
+func LateInitializeNodeGroup(in *v1alpha1.NodeGroupParameters, ng *eks.Nodegroup) { // nolint:gocyclo
+	if ng == nil {
+		return
+	}
+	in.AMIType = v1alpha1.AMITypes(awsclients.LateInitializeString(string(in.AMIType), awsclients.String(string(ng.AmiType))))
+	in.DiskSize = awsclients.LateInitializeInt64Ptr(in.DiskSize, ng.DiskSize)
+	if len(in.InstanceTypes) == 0 && len(ng.InstanceTypes) > 0 {
+		in.InstanceTypes = ng.InstanceTypes
+	}
+	if len(in.Labels) == 0 && len(ng.Labels) > 0 {
+		in.Labels = ng.Labels
+	}
+	if in.RemoteAccess == nil && ng.RemoteAccess != nil {
+		in.RemoteAccess = &v1alpha1.RemoteAccessConfig{
+			EC2SSHKey:            ng.RemoteAccess.Ec2SshKey,
+			SourceSecurityGroups: ng.RemoteAccess.SourceSecurityGroups,
+		}
+	}
+	if in.ScalingConfig == nil && ng.ScalingConfig != nil {
+		in.ScalingConfig = &v1alpha1.NodeGroupScalingConfig{
+			DesiredSize: ng.ScalingConfig.DesiredSize,
+			MinSize:     ng.ScalingConfig.MinSize,
+			MaxSize:     ng.ScalingConfig.MaxSize,
+		}
+	}
+	in.ReleaseVersion = awsclients.LateInitializeStringPtr(in.ReleaseVersion, ng.ReleaseVersion)
+	in.Version = awsclients.LateInitializeStringPtr(in.Version, ng.Version)
+	// NOTE(hasheddan): we always will set the default Crossplane tags in
+	// practice during initialization in the controller, but we check if no tags
+	// exist for consistency with expected late initialization behavior.
+	if in.Tags == nil {
```

Comment:
```suggestion
	if len(in.Tags) == 0 {
```

provides a more guaranteed result.